lib/multigrid.cpp

```diff
@@ -187,32 +177,35 @@ namespace quda {
       setOutputPrefix(prefix); // restore since we just popped back from coarse grid
 
       createCoarseSolver();
+    }
 
+    if (param.level == 0) {
       // now we can run the verification if requested
       if (param.mg_global.run_verify) verify();
-
-      // resize the on-GPU null-space components to single-parity if we're doing a
-      // single-parity solve (memory saving technique).
-      {
-        QudaSiteSubset site_subset = param.coarse_grid_solution_type == QUDA_MATPC_SOLUTION ? QUDA_PARITY_SITE_SUBSET : QUDA_FULL_SITE_SUBSET;
-        QudaMatPCType matpc_type = param.mg_global.invert_param->matpc_type;
-        QudaParity parity = (matpc_type == QUDA_MATPC_EVEN_EVEN || matpc_type == QUDA_MATPC_EVEN_EVEN_ASYMMETRIC) ? QUDA_EVEN_PARITY : QUDA_ODD_PARITY;
-        transfer->setSiteSubset(site_subset, parity); // use this to force location of transfer
-      }
     }
 
     if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Setup of level %d of %d done\n", param.level+1, param.Nlevel);
 
-    // print out profiling information for the adaptive setup
-    if (getVerbosity() >= QUDA_VERBOSE) profile.Print();
-    // Reset the profile for accurate solver timing
-    profile.TPRESET();
+    popLevel(param.level);
+  }
 
+  void MG::pushLevel(int level) const
+  {
     postTrace();
+    pushVerbosity(param.mg_global.verbosity[level]);
+    setOutputPrefix(prefix);
   }
 
-  void MG::destroySmoother() {
+  void MG::popLevel(int level) const
+  {
+    popVerbosity();
+    setOutputPrefix("");
```

Comment:
Why does this get set to the empty string instead of the previous output prefix? Should we have a `pushOutputPrefix`/`popOutputPrefix` for MG?

Reply:
Well this was consistent with prior behaviour, but I'll add new `pushOutputPrefix` and `popOutputPrefix` functions to address this and simplify things.